### PR TITLE
backend/drm: Retry disable without unsetting mode and crtc id

### PR DESCRIPTION
**Pull request by agx**
otherwise we can't blank/unblank on DSI with fixed pipeline.

Closes: #1889

### CODE REVIEW

**Comment by emersion** [comment]
We really want to unset the CRTC, otherwise it keeps the pipeline on and we can't use it for another connector.

**Comment by agx** [comment]
@ascent12 can you point me to the uapi parts that state it should work? With the tracing i did in the kernel it seemed to not even hit the display driver (let alone lower layers like panel or dsi host controller)  upon tht request but rejects things in generic drm already, see also the debug output in #1889

**Comment by agx** [comment]
this looks pretty different now and works for me on arm64 (mxsfb) and intel.

File: backend/drm/atomic.c [hunk 1]
```diff
@@ -140,16 +140,19 @@ static bool atomic_conn_enable(struct wlr_drm_backend *drm,
 		return !enable;
 	}
 
+	/* crtc will be enabled on next page flip */
+	if (enable) {
+		crtc->needs_modeset = true;
```
Comment (emersion) [inline backend/drm/atomic.c]: This seems to set `DRM_MODE_ATOMIC_ALLOW_MODESET`. However we already set it in the `atomic_commit` call, so I don't think it's necessary.

File: backend/drm/atomic.c [hunk 2]
```diff
@@ -140,16 +140,19 @@ static bool atomic_conn_enable(struct wlr_drm_backend *drm,
...
 	atomic_begin(crtc, &atom);
 	atomic_add(&atom, crtc->id, crtc->props.active, enable);
-	if (enable) {
-		atomic_add(&atom, conn->id, conn->props.crtc_id, crtc->id);
-		atomic_add(&atom, crtc->id, crtc->props.mode_id, crtc->mode_id);
-	} else {
-		atomic_add(&atom, conn->id, conn->props.crtc_id, 0);
-		atomic_add(&atom, crtc->id, crtc->props.mode_id, 0);
-	}
+	atomic_add(&atom, conn->id, conn->props.crtc_id, 0);
```
Comment (emersion) [inline backend/drm/atomic.c]: This disables the plane even if `enable == true`.

Comment (agx) [reply backend/drm/atomic.c]: no, we return in early in the `enable` case, see above.

Comment (emersion) [reply backend/drm/atomic.c]: Hmm. So we don't commit anymore. If we don't commit, we don't know whether we can enable the output (bandwidth limits).

File: backend/drm/atomic.c (hunk 1 above)
Comment (agx) [reply backend/drm/atomic.c]: it's for the case where we `enable` the crtc but `atomic_crtc_pageflip()` does not get passed a mode.

File: backend/drm/atomic.c (hunk 2 above)
Comment (agx) [reply backend/drm/atomic.c]: but isn't this only really useful if we also assign plane props, etc. which happens on pageflip (which is where it happens now)?

**Comment by agx** [comment]
Seems kmscube and weston use same pattern of putting the next commit instead of handling it separateley as wlroots currently does:

https://gitlab.freedesktop.org/wayland/weston/blob/master/libweston%2Fbackend-drm%2Fkms.c#L972

This allows to have this together with the fb properties which is needed if we unassign that on disable (which seems needed for drivers that use `drm_simple_kms_crtc_check` (see #1889).

File: backend/drm/atomic.c (hunk 2 above)
Comment (agx) [reply backend/drm/atomic.c]: Is there another solution i'm missing? it'd be nice to have these drm drivers also work out of the box with wlroots.

**Comment by emersion** [comment]
This should be fixed by https://github.com/swaywm/wlroots/pull/2169 as well.

**Comment by emersion** [comment]
@agx Is this PR still required with the new DRM backend changes that landed in master?

**Comment by agx** [comment]
ah...did not see that ping back in may. I'll put checking that on my todo list (currently basing on 0.10.1 here)

**Comment by emersion** [comment]
Closing because I think it's fixed.

**Pull request closed without merging**